### ISSUE TITLE
Use `ubuntu:20.04` vs `ubuntu:20.10` for base image?

### ISSUE THREAD

**Issue by joshbranham** [argoproj/argo-cd#8763]
# Summary

Ubuntu 20.10 is EOL per https://wiki.ubuntu.com/Releases, which means if a user tries to build a custom ArgoCD image from your base, they cannot install packages etc.

# Motivation

We build a custom ArgoCD image from `quay.io/argoproj/argocd:v2.0.5` to add our custom plugin. Trying to upgrade to 2.x of ArgoCD revealed this issue.

```shell
❯ docker run --rm -it ubuntu:20.10 bash
root@3a8838274293:/# apt-get update
Ign:1 http://ports.ubuntu.com/ubuntu-ports groovy InRelease
Ign:2 http://ports.ubuntu.com/ubuntu-ports groovy-updates InRelease
Ign:3 http://ports.ubuntu.com/ubuntu-ports groovy-backports InRelease
Ign:4 http://ports.ubuntu.com/ubuntu-ports groovy-security InRelease
Err:5 http://ports.ubuntu.com/ubuntu-ports groovy Release
  404  Not Found [IP: 91.189.88.152 80]
Err:6 http://ports.ubuntu.com/ubuntu-ports groovy-updates Release
  404  Not Found [IP: 91.189.88.152 80]
Err:7 http://ports.ubuntu.com/ubuntu-ports groovy-backports Release
  404  Not Found [IP: 91.189.88.152 80]
Err:8 http://ports.ubuntu.com/ubuntu-ports groovy-security Release
  404  Not Found [IP: 91.189.88.152 80]
Reading package lists... Done
E: The repository 'http://ports.ubuntu.com/ubuntu-ports groovy Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://ports.ubuntu.com/ubuntu-ports groovy-updates Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://ports.ubuntu.com/ubuntu-ports groovy-backports Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://ports.ubuntu.com/ubuntu-ports groovy-security Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
root@3a8838274293:/#
```

# Proposal

Modify the base image to use an LTS version of Ubuntu `20.04` with supported apt repos.

>Note: It is possible I am missing something else, so I apologize if that is the case, but I am surprised no one else is running into this. I would prefer not to fork and build my own image from your Dockerfile if at all possible.

**Comment by joshbranham** [comment]
Looks like it has been changed to `21.10` in a newer version which begs the question:

- What is the preferred upgrade path for folks going from 1.x versions to 2.x, where the base image in the not-latest 2.x releases is broken?
- Why not use LTS? I was hoping to see more elaboration on this comment/PR on the decision to switch to Ubuntu and opt against LTS https://github.com/argoproj/argo-cd/issues/5047#issuecomment-752719328

**Comment by jannfis** [comment]
I'm at the point where I don't oppose considering an LTS base image for our builds.

I guess the main reason for not going LTS was, that they didn't fix medium and low security issues, at least in my initial tests at https://github.com/argoproj/argo-cd/issues/5047#issuecomment-744442270

**Comment by joshbranham** [comment]
> I'm at the point where I don't oppose considering an LTS base image for our builds.
> 
> I guess the main reason for not going LTS was, that they didn't fix medium and low security issues, at least in my initial tests at [#5047 (comment)](https://github.com/argoproj/argo-cd/issues/5047#issuecomment-744442270)

Is running `apt-get upgrade` in the `argo-cd` image build process considered too aggressive? I would assume most CVEs would be patched through that, coupled with a solid test suite of the resulting image to ensure any dependent packages on the OS work as expected.

**Comment by crenshaw-dev** [comment]
@joshbranham we run `apt-get dist-upgrade` in the build, which I think for the purposes of this discussion is functionally equivalent to `apt-get upgrade`.

**Comment by crenshaw-dev** [comment]
Just ran a no-cache build locally with both bases. 21.10 wins by a bit.

| | 20.04 vs 21.10 | 
|-|-|
| High | N/A |
| Medium | +2/-0 |
| Low | +5/-2 |
| Total | +7/-2 |

idk. I don't feel strongly either way.

**Comment by crenshaw-dev** [comment]
Oh. To respond to Josh's specific question about the base image for Argo CD 2.0.x. We're only doing patch releases for the three most recent major releases. The current base image for 2.1.x is 21.04 which is not yet EOL, so I'd recommend upgrading. https://github.com/argoproj/argo-cd/blob/273a952e6c03c3a5e6aaa666a00a3e275128776a/Dockerfile#L1

**Comment by joshbranham** [comment]
> Oh. To respond to Josh's specific question about the base image for Argo CD 2.0.x. We're only doing patch releases for the three most recent major releases. The current base image for 2.1.x is 21.04 which is not yet EOL, so I'd recommend upgrading.
> 
> https://github.com/argoproj/argo-cd/blob/273a952e6c03c3a5e6aaa666a00a3e275128776a/Dockerfile#L1

Yeah I landed on `2.2.7` but we were on an older `1.x` release so I was trying to upgrade in steps and check the changelog for breaking changes as I went, and in our case, we build a custom image so it broke on 2.0.x I think. Sounds like if one keeps up to date with the newer releases for any major + minor then it is a non-issue. With that I will close this out, thanks for getting back to me.
